### PR TITLE
Create converters.py containing functions for converting PrivacyPayload

### PR DESCRIPTION
**Pull request by sonhal**
objects to pandas.DataFrame

Create state_printer.py containing functions for printing AaaS state

Create tests for converter.py functions

Create aaas.describe() method for providing information on the current state

### CODE REVIEW

File: pyaaas/aaas.py [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+from collections.abc import MutableMapping, Sequence, Callable
 import json
 
 from pyaaas.aaas_connector import AaaSConnector
 from pyaaas.models.anonymize_payload import AnonymizePayload
 from pyaaas.models.privacy_models import PrivacyModel
 from pyaaas.models.anonymize_result import AnonymizeResult
+from pyaaas.state_printer import jupyter_print_mapping, print_mapping
+from pyaaas.converters import create_privacy_models_dataframe,create_attribute_types_dataframe,\
+    create_transform_models_dataframe
```
Comment (JeremiahUy) [inline pyaaas/aaas.py:13]: Remove whitespace please

File: pyaaas/aaas.py [hunk 2]
```diff
@@ -26,6 +29,11 @@ def __init__(self, url: str=None):
 
         self._payload = AnonymizePayload()
```
Comment (JeremiahUy) [inline pyaaas/aaas.py:31]: remove whitespace please

File: pyaaas/converters.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+from collections.abc import Sequence, Mapping
+
+import pandas
+
```
Comment (JeremiahUy) [inline pyaaas/converters.py:4]: remove whitespace please

File: pyaaas/converters.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        model_index =[model.name]
+        for key, value in model.items():
+            model_index.append(key)
+            privacy_models_values.append(value)
+        privacy_models_index.append(model_index)
+
+    index = pandas.MultiIndex.from_tuples(privacy_models_index, names=("privacy_model", "parameter"))
+    dataframe = pandas.DataFrame(privacy_models_values, index=index, columns=("value",))
+    return dataframe
+
```
Comment (JeremiahUy) [inline pyaaas/converters.py:26]: remove whitespace please

Comment (JeremiahUy) [inline pyaaas/converters.py]: Remove whitespace please

File: pyaaas/converters.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+def create_attribute_types_dataframe(attribute_types: Mapping) -> pandas.DataFrame:
+    """
+    Creates a pandas.Dataframe from a Mapping of attribute_types str:field=str:attribute_type
+
+    :param attribute_types: Mapping str:field=str:attribute_type
+    :return: pandas.DataFrame
+    """
+    attribute_rows = attribute_types.items()
+    return pandas.DataFrame(attribute_rows, columns=("field", "type")).set_index("field")
+
```
Comment (JeremiahUy) [inline pyaaas/converters.py:37]: remove whitespace please

File: pyaaas/converters.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+def create_transform_models_dataframe(transform_models):
+    """
+    Creates a pandas.DataFrame form a Mapping of Transform Models
+
```
Comment (JeremiahUy) [inline pyaaas/converters.py:42]: remove whitespace please

File: pyaaas/converters.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    for field, hierarchy in transform_models.items():
+        field_index, field_values = _create_index_and_values_for(field, hierarchy)
+        transform_models_index += field_index
+        transform_models_values += field_values
+
+    index = pandas.MultiIndex.from_tuples(transform_models_index)
+    dataframe = pandas.DataFrame(transform_models_values, index=index)
+    return dataframe
+
```
Comment (JeremiahUy) [inline pyaaas/converters.py:58]: remove whitespace please

File: pyaaas/converters.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    for level in range(1, _get_hierarchy_levels(hierarchy)):
+        model_index = [field]
+        level_values = []
+        model_index.append(f"level_{level}")
+        transform_model_index.append(model_index)
+        for row in hierarchy:
+            level_values.append(row[level])
+        transform_model_values.append(level_values)
+    return transform_model_index, transform_model_values
+
```
Comment (JeremiahUy) [inline pyaaas/converters.py:73]: remove whitespace please

File: pyaaas/state_printer.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
+from collections.abc import Mapping, Callable
+from IPython.core import display as ICD
+import pandas
+
```
Comment (JeremiahUy) [inline pyaaas/state_printer.py:4]: remove whitespace please

File: pyaaas/state_printer.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    Prints out a human readable view of the provided Privacy Models
+
+    :param models: collection of Privacy models
+    :param printer: callable
+    :return: None
+    """
+    for model in models.values():
+        return _model_to_dataframe(model)
+
```
Comment (JeremiahUy) [inline pyaaas/state_printer.py:16]: remove whitespace please

File: pyaaas/state_printer.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    :param name_data_mapping: Mapping object name:data
+    :return: None
+    """
+
+    for name, dataframe in name_data_mapping.items():
+        print(name)
+        print(dataframe)
+        print("-"*40 + "\n")
+
```
Comment (JeremiahUy) [inline pyaaas/state_printer.py:30]: remove whitespace please

File: pyaaas/state_printer.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    Print the provided name_dataframe_mapping in a J
+
+    :param name_dataframe_mapping:
+    :return:
+    """
+    for name, dataframe in name_dataframe_mapping.items():
+        print(name)
+        ICD.display(dataframe)
+        print("-"*40)
+
```
Comment (JeremiahUy) [inline pyaaas/state_printer.py:43]: remove whitespace please

File: pyaaas/state_printer.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+def print_attribute_types(attributes, printer):
+    for attribute in attributes:
+        printer(attribute)
+
```
Comment (JeremiahUy) [inline pyaaas/state_printer.py]: remove whitespace please

File: tests/pyaaas/test_converters.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
+import unittest
+
```
Comment (JeremiahUy) [inline tests/pyaaas/test_converters.py:2]: remove whitespace please

File: tests/pyaaas/test_converters.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import pandas
+
+from pyaaas import converters
+from pyaaas.models.privacy_models import KAnonymity
+
```
Comment (JeremiahUy) [inline tests/pyaaas/test_converters.py:7]: remove whitespace please

File: tests/pyaaas/test_converters.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        self.test_privacy_models_sequence = [KAnonymity(k=4)]
+        self.test_attribute_types_mapping = {"id": "INDENTIFING",
+                                             "name": "QUASIIDENTIFING"}
+
+    def test_create_transform_models_dataframe__run(self):
+        dataframe = converters.create_transform_models_dataframe(self.test_hierarchy_mapping)
+        self.assertIsNotNone(dataframe)
+        self.assertIsInstance(dataframe, pandas.DataFrame)
+
```
Comment (JeremiahUy) [inline tests/pyaaas/test_converters.py:32]: remove whitespace please

File: tests/pyaaas/test_state_printer.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
+import unittest
+import builtins
+
```
Comment (JeremiahUy) [inline tests/pyaaas/test_state_printer.py:3]: remove whitespace please

File: tests/pyaaas/test_state_printer.py [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+import pandas
+
+from pyaaas.aaas import AaaS
+from pyaaas.models.privacy_models import KAnonymity
+from pyaaas import state_printer
+
```
Comment (JeremiahUy) [inline tests/pyaaas/test_state_printer.py:7]: remove whitespace please

File: tests/pyaaas/test_state_printer.py [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+def mock_print(*args):
+    pass
+
```
Comment (JeremiahUy) [inline tests/pyaaas/test_state_printer.py:11]: remove whitespace please

File: tests/pyaaas/test_state_printer.py [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+
+builtins.print = mock_print # mock print function
+
```
Comment (JeremiahUy) [inline tests/pyaaas/test_state_printer.py:14]: remove whitespace please

File: tests/pyaaas/test_state_printer.py [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    def setUp(self):
+
+        self.test_df = pandas.DataFrame()
+        self.test_name_dataframe_mapping ={"zipcode": pandas.DataFrame()}
+
+    def test_print_mapping__run(self):
+        state_printer.print_mapping(self.test_name_dataframe_mapping)
+
+
```
Comment (JeremiahUy) [inline tests/pyaaas/test_state_printer.py:28]: remove whitespace please

**Review by JeremiahUy — DISMISSED** [review]
remove whitespaces on commented lines :D

**Review by JulianSagen — APPROVED** [review]
Neat, nice and pretty straight forward.
10/10 Approved

**Review by JeremiahUy — APPROVED** [review]

**Pull request merged**